super(self.__class__, self) causes INTERNAL ERROR (AssertionError: unreachable)

This is a bug.
<br>**Reproducer**
```python
import types
from typing import TypeVar

class Foo:
    def f(self):
        print('Foo')

class Bar(Foo):
    def f(self):
        print('Bar')

T = TypeVar('T')

def my_f(self: T) -> None:
    super(self.__class__, self).f()
    print('monkeypatch')

bar = Bar()
bar.f = types.MethodType(my_f, bar)
bar.f()
```
This code may not be perfect, but it is accepted by the python interpreter.
<br>**Actual Behavior**
```
repr.py:15: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.rtfd.io/en/latest/common_issues.html#using-a-development-mypy-build
If this issue continues with mypy master, please report a bug at https://github.com/python/mypy/issues
version: 0.770+dev.150d4921e702512e8afa0fc9cc22e71e720c3069
repr.py:15: : note: please use --show-traceback to print a traceback when reporting a bug
```
<br>**Expected Behavior**
mypy does not crash.
<br>**Versions**
mypy: Reproducible on 0.761 release and git master (commit 150d492)
python: 3.8.1
<br>**Flags Used**
No flags are necessary.
<br>**Traceback**
```
Traceback (most recent call last):
  File "/usr/bin/mypy", line 11, in <module>
    load_entry_point('mypy==0.770+dev.150d4921e702512e8afa0fc9cc22e71e720c3069', 'console_scripts', 'mypy')()
  File "/usr/lib/python3.8/site-packages/mypy/__main__.py", line 8, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "/usr/lib/python3.8/site-packages/mypy/main.py", line 89, in main
    res = build.build(sources, options, None, flush_errors, fscache, stdout, stderr)
  File "/usr/lib/python3.8/site-packages/mypy/build.py", line 165, in build
    result = _build(
  File "/usr/lib/python3.8/site-packages/mypy/build.py", line 234, in _build
    graph = dispatch(sources, manager, stdout)
  File "/usr/lib/python3.8/site-packages/mypy/build.py", line 2631, in dispatch
    process_graph(graph, manager)
  File "/usr/lib/python3.8/site-packages/mypy/build.py", line 2938, in process_graph
    process_stale_scc(graph, scc, manager)
  File "/usr/lib/python3.8/site-packages/mypy/build.py", line 3037, in process_stale_scc
    graph[id].type_check_first_pass()
  File "/usr/lib/python3.8/site-packages/mypy/build.py", line 2130, in type_check_first_pass
    self.type_checker().check_first_pass()
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 293, in check_first_pass
    self.accept(d)
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 400, in accept
    stmt.accept(self)
  File "/usr/lib/python3.8/site-packages/mypy/nodes.py", line 676, in accept
    return visitor.visit_func_def(self)
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 725, in visit_func_def
    self._visit_func_def(defn)
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 729, in _visit_func_def
    self.check_func_item(defn, name=defn.name)
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 791, in check_func_item
    self.check_func_def(defn, typ, name)
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 974, in check_func_def
    self.accept(item.body)
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 400, in accept
    stmt.accept(self)
  File "/usr/lib/python3.8/site-packages/mypy/nodes.py", line 1004, in accept
    return visitor.visit_block(self)
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 1969, in visit_block
    self.accept(s)
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 400, in accept
    stmt.accept(self)
  File "/usr/lib/python3.8/site-packages/mypy/nodes.py", line 1019, in accept
    return visitor.visit_expression_stmt(self)
  File "/usr/lib/python3.8/site-packages/mypy/checker.py", line 3089, in visit_expression_stmt
    self.expr_checker.accept(s.expr, allow_none_return=True, always_allow_any=True)
  File "/usr/lib/python3.8/site-packages/mypy/checkexpr.py", line 3745, in accept
    typ = self.visit_call_expr(node, allow_none_return=True)
  File "/usr/lib/python3.8/site-packages/mypy/checkexpr.py", line 263, in visit_call_expr
    return self.visit_call_expr_inner(e, allow_none_return=allow_none_return)
  File "/usr/lib/python3.8/site-packages/mypy/checkexpr.py", line 312, in visit_call_expr_inner
    callee_type = get_proper_type(self.accept(e.callee, type_context, always_allow_any=True))
  File "/usr/lib/python3.8/site-packages/mypy/checkexpr.py", line 3749, in accept
    typ = node.accept(self)
  File "/usr/lib/python3.8/site-packages/mypy/nodes.py", line 1834, in accept
    return visitor.visit_super_expr(self)
  File "/usr/lib/python3.8/site-packages/mypy/checkexpr.py", line 3474, in visit_super_expr
    assert False, 'unreachable'
AssertionError: unreachable
```

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -3464,6 +3464,10 @@ class ExpressionChecker(ExpressionVisitor[Type]):
             self.chk.fail(message_registry.SUPER_ARG_2_NOT_INSTANCE_OF_ARG_1, e)
             return AnyType(TypeOfAny.from_error)
 
+        if len(mro) == index + 1:
+            self.chk.fail(message_registry.SUPER_OUTSIDE_OF_METHOD_NOT_SUPPORTED, e)
+            return AnyType(TypeOfAny.from_error)
+
         for base in mro[index+1:]:
             if e.name in base.names or base == mro[-1]:
                 if e.info and e.info.fallback_to_any and base == mro[-1]:

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -3465,7 +3465,7 @@ class ExpressionChecker(ExpressionVisitor[Type]):
             return AnyType(TypeOfAny.from_error)
 
         if len(mro) == index + 1:
-            self.chk.fail(message_registry.SUPER_OUTSIDE_OF_METHOD_NOT_SUPPORTED, e)
+            self.chk.fail(message_registry.TARGET_CLASS_HAS_NO_BASE_CLASS, e)
             return AnyType(TypeOfAny.from_error)
 
         for base in mro[index+1:]:

--- a/mypy/message_registry.py
+++ b/mypy/message_registry.py
@@ -109,6 +109,7 @@ SUPER_VARARGS_NOT_SUPPORTED = 'Varargs not supported with "super"'  # type: Fina
 SUPER_POSITIONAL_ARGS_REQUIRED = '"super" only accepts positional arguments'  # type: Final
 SUPER_ARG_2_NOT_INSTANCE_OF_ARG_1 = \
     'Argument 2 for "super" not an instance of argument 1'  # type: Final
+TARGET_CLASS_HAS_NO_BASE_CLASS = 'Target class has no base class'  # type: Final
 SUPER_OUTSIDE_OF_METHOD_NOT_SUPPORTED = \
     'super() outside of a method is not supported'  # type: Final
 SUPER_ENCLOSING_POSITIONAL_ARGS_REQUIRED = \

--- a/test-data/unit/check-super.test
+++ b/test-data/unit/check-super.test
@@ -280,6 +280,20 @@ class B(A):
 class C:
     a = super().whatever  # E: super() outside of a method is not supported
 
+[case testSuperWithObjectClassAsFirstArgument]
+class A:
+    def f(self) -> None:
+        super(object, self).f()  # E: super() outside of a method is not supported
+
+[case testSuperWithTypeVarAsFirstArgument]
+from typing import TypeVar
+
+T = TypeVar('T')
+
+def f(obj: T) -> None:
+    super(obj.__class__, obj).f()  # E: super() outside of a method is not supported
+[builtins fixtures/__new__.pyi]
+
 [case testSuperWithSingleArgument]
 class B:
     def f(self) -> None: pass

--- a/test-data/unit/check-super.test
+++ b/test-data/unit/check-super.test
@@ -283,7 +283,7 @@ class C:
 [case testSuperWithObjectClassAsFirstArgument]
 class A:
     def f(self) -> None:
-        super(object, self).f()  # E: super() outside of a method is not supported
+        super(object, self).f()  # E: Target class has no base class
 
 [case testSuperWithTypeVarAsFirstArgument]
 from typing import TypeVar
@@ -291,7 +291,7 @@ from typing import TypeVar
 T = TypeVar('T')
 
 def f(obj: T) -> None:
-    super(obj.__class__, obj).f()  # E: super() outside of a method is not supported
+    super(obj.__class__, obj).f()  # E: Target class has no base class
 [builtins fixtures/__new__.pyi]
 
 [case testSuperWithSingleArgument]

--- a/test-data/unit/fixtures/__new__.pyi
+++ b/test-data/unit/fixtures/__new__.pyi
@@ -5,6 +5,8 @@ from typing import Any
 class object:
     def __init__(self) -> None: pass
 
+    __class__ = object
+
     def __new__(cls) -> Any: pass
 
 class type:

